### PR TITLE
Add generic AutoPopStack class

### PR DESCRIPTION
**Pull request by ldalek-antmicro**
Generic autopop stack class is intended to be used in code base which uses stack-based algorithms like symbol or node tracking, e.g. `SyntaxTreeContext`.

### CODE REVIEW

File: common/util/auto_pop_stack.h [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+  const value_type& top() const {
+    CHECK(!stack_.empty());
+    return stack_.back();
+  }
+
+  // All elements of stack are non-null because they refer to traversed nodes.
+  typedef std::vector<value_type> stack_type;
+  typedef typename stack_type::iterator iterator;
+  typedef typename stack_type::const_iterator const_iterator;
+  typedef typename stack_type::const_reverse_iterator const_reverse_iterator;
```
Comment (fangism) [inline common/util/auto_pop_stack.h:68]: Move these typedefs near top of class.

File: common/util/auto_pop_stack.h [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  // Reverse iterators be useful for searching from the top-of-stack downward.
+  const_reverse_iterator rbegin() const { return stack_.rbegin(); }
+  const_reverse_iterator rend() const { return stack_.rend(); }
+
+ protected:
+  value_type& top() {
+    // Efficient C++, 3rd, Item 3, "Use const whenever possible."
+    return const_cast<value_type&>(
+        static_cast<const this_type&>(*this).top());
+  }
```
Comment (fangism) [inline common/util/auto_pop_stack.h:83]: No need for casting, just `return stack_.back();`.  This should internally look identical to the `const` method counterpart.

File: common/util/auto_pop_stack.h [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+  // Push a value_type onto the stack (takes r-value reference)
+  void Push(value_type&& value) {
+    stack_.push_back(std::move(value));
+  }
+
+  // Push a value_type onto the stack (takes reference)
+  void Push(const basic_value_type& value) {
```
Comment (fangism) [inline common/util/auto_pop_stack.h:91]: `const value_type&`

File: common/util/auto_pop_stack.h [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+      stack->Push(std::move(value));
+    }
+    AutoPop(this_type* stack, basic_value_type& value) {
+      stack_ = stack;
+      stack->Push(&value);
+    }
+    ~AutoPop() {
+      stack_->Pop();
+    }
+
```
Comment (fangism) [inline common/util/auto_pop_stack.h:47]: Suggest also deleting many standard interfaces, as this is intended for use as a stack object.

```C++
AutoPop& operator=(const AutoPop&) = delete;  // copy-assign
AutoPop& operator=(AutoPop&&) = delete;  // move-assign
AutoPop(const AutoPop&) = delete;  // copy-construct
AutoPop(AutoPop&&) = delete;  // move-construct
```

File: common/util/auto_pop_stack.h [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+#include "common/util/iterator_adaptors.h"
+
+namespace verible {
+
+// AutoPop stack class. Templated version of SyntaxTreeContext::AutoPop
+template <typename T>
+class AutoPopStack {
+ public:
+  typedef T value_type;
+  typedef typename std::remove_pointer<value_type>::type basic_value_type;
```
[the remark this answers is not in the document]
Comment (ldalek-antmicro) [reply common/util/auto_pop_stack.h]: Done & done

File: common/util/auto_pop_stack.h [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+  typedef AutoPopStack<value_type> this_type;
+
+  // member class to handle push and pop of stack safely
+  class AutoPop {
+   public:
+    AutoPop(this_type* stack, value_type&& value) {
+      stack_ = stack;
+      stack->Push(std::move(value));
+    }
+    AutoPop(this_type* stack, basic_value_type& value) {
```
[the remark this answers is not in the document]
Comment (ldalek-antmicro) [reply common/util/auto_pop_stack.h]: Done

File: common/util/auto_pop_stack.h [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+// AutoPop stack class. Templated version of SyntaxTreeContext::AutoPop
+template <typename T>
+class AutoPopStack {
+ public:
+  typedef T value_type;
+  typedef typename std::remove_pointer<value_type>::type basic_value_type;
+  typedef AutoPopStack<value_type> this_type;
+
+  // member class to handle push and pop of stack safely
```
[the remark this answers is not in the document]
Comment (ldalek-antmicro) [reply common/util/auto_pop_stack.h]: Done

File: common/util/auto_pop_stack.h (hunk 1 above)
Comment (ldalek-antmicro) [reply common/util/auto_pop_stack.h]: Done

File: common/util/auto_pop_stack.h (hunk 2 above)
Comment (ldalek-antmicro) [reply common/util/auto_pop_stack.h]: Just wanted to reuse `const` `top()` method. Fixed.

File: common/util/auto_pop_stack.h (hunk 3 above)
Comment (ldalek-antmicro) [reply common/util/auto_pop_stack.h]: Done

File: common/util/auto_pop_stack.h (hunk 4 above)
Comment (ldalek-antmicro) [reply common/util/auto_pop_stack.h:55]: Done

File: common/text/syntax_tree_context.h [hunk 8]
```diff
@@ -124,26 +104,21 @@ class SyntaxTreeContext {
...
-  // Push a SyntaxTreeNode onto the stack (takes address)
-  void Push(const SyntaxTreeNode& node);
-
   // Stack of ancestors of the current node that is updated as the tree
   // is traversed. Top of the stack is closest ancestor.
   // A vector is chosen to allow random access and searches from either end of
   // the stack.
-  stack_type stack_;
+  // private:
+  //  base_type::stack_type base_type::stack_;
```
Comment (fangism) [inline common/text/syntax_tree_context.h:121]: Can delete this section now.

File: common/util/auto_pop_stack.h [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef VERIBLE_COMMON_UTIL_AUTO_POP_STACK_H_
+#define VERIBLE_COMMON_UTIL_AUTO_POP_STACK_H_
+
+#include <iterator>
+#include <type_traits>
+#include <vector>
+#include "common/util/iterator_adaptors.h"
```
Comment (fangism) [inline common/util/auto_pop_stack.h:21]: minor nit: blank line above here to separate groups of headers

File: common/util/auto_pop_stack.h [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+namespace verible {
+
+// AutoPop stack class. Templated version of SyntaxTreeContext::AutoPop
```
Comment (fangism) [inline common/util/auto_pop_stack.h:25]: Update comment to not mention SyntaxTreeContext::AutoPop (reference is cyclic).
Suggest describing use cases of this class, and a motivating example.
Also worth mentioning that this provides public read-only random access/iteration into the stack elements.

File: common/util/auto_pop_stack.h [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+template <typename T>
+class AutoPopStack {
+ public:
+  typedef T value_type;
+  typedef AutoPopStack<value_type> this_type;
+
+  typedef value_type& reference_type;
+  typedef const value_type& const_reference_type;
```
Comment (fangism) [inline common/util/auto_pop_stack.h:33]: Let's follow STL convention and call this `const_reference`.

File: common/util/auto_pop_stack.h [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef VERIBLE_COMMON_UTIL_AUTO_POP_STACK_H_
+#define VERIBLE_COMMON_UTIL_AUTO_POP_STACK_H_
+
+#include <iterator>
```
Comment (fangism) [inline common/util/auto_pop_stack.h:18]: add (for `size_t`):

#include <cstddef>

File: common/text/syntax_tree_context.h [hunk 13]
```diff
@@ -31,38 +33,16 @@ namespace verible {
 // Note: Public methods are named to follow STL convention for std::stack.
 // TODO(fangism): implement a general ForwardMatcher and ReverseMatcher
 // interface that can express AND/OR/NOT.
-class SyntaxTreeContext {
+class SyntaxTreeContext : public AutoPopStack<const SyntaxTreeNode*> {
```
Comment (fangism) [inline common/text/syntax_tree_context.h:36]: Document that one of the invariants is that elements of the (internal) stack are restricted to non-nullptrs.  (This was the case before refactoring.)

File: common/text/syntax_tree_context.h [hunk 14]
```diff
@@ -31,38 +33,16 @@ namespace verible {
...
-
-  // Allow read-only random-access into stack:
-  const_iterator begin() const { return stack_.begin(); }
-  const_iterator end() const { return stack_.end(); }
-
-  // Reverse iterators be useful for searching from the top-of-stack downward.
-  const_reverse_iterator rbegin() const { return stack_.rbegin(); }
-  const_reverse_iterator rend() const { return stack_.rend(); }
+  const SyntaxTreeNode& top() const {
+    return *ABSL_DIE_IF_NULL(base_type::top());
```
Comment (fangism) [inline common/text/syntax_tree_context.h:44]: You can hide the base class's interface with:

```c++
protected:
  using base_type::top;
```

This makes it clear that you are further restricting direct uses of the base class methods intentionally, and here is where you can alter the interface to return references if you wish.  

Optionally, you can also be explicit about what base class interfaces you wish to expose:

```c++
public:
  using base_type::begin;
```

File: common/text/syntax_tree_context.h (hunk 8 above)
Comment (ldalek-antmicro) [reply common/text/syntax_tree_context.h]: Done

File: common/util/auto_pop_stack.h (hunk 9 above)
Comment (ldalek-antmicro) [reply common/util/auto_pop_stack.h:23]: Done

File: common/util/auto_pop_stack.h (hunk 10 above)
Comment (ldalek-antmicro) [reply common/util/auto_pop_stack.h]: Done

File: common/util/auto_pop_stack.h (hunk 11 above)
Comment (ldalek-antmicro) [reply common/util/auto_pop_stack.h]: Also fixed `reference`

File: common/util/auto_pop_stack.h (hunk 12 above)
Comment (ldalek-antmicro) [reply common/util/auto_pop_stack.h:19]: Added `cstddef`. Hope it's that what you meant.

File: common/text/syntax_tree_context.h (hunk 13 above)
Comment (ldalek-antmicro) [reply common/text/syntax_tree_context.h:38]: Done

File: common/text/syntax_tree_context.h (hunk 14 above)
Comment (ldalek-antmicro) [reply common/text/syntax_tree_context.h:52]: Just restricted access to `base_type::top`. Also removed `base_type::` from methods like `rbegin`, `rend` and so on.

File: common/util/auto_pop_stack.h (hunk 12 above)
Comment (fangism) [reply common/util/auto_pop_stack.h:19]: yes.

**Pull request closed without merging**